File: EPI/17.5/swift/17.5.swift
private typealias Cell = (i: Int, j: Int)

public func contains(grid: [[Int]], sequence: [Int]) -> Bool {
    if sequence.isEmpty {
        return true
    }

    // impossible[i][j] contains k if there is no path starting at (i, j) that retraces the sequence
    // suffix of length k.
    var impossible = grid.map { row in
        row.map { _ in Set<Int>() }
    }

    return grid.enumerated().contains { (i, row) in
        row.enumerated().contains { (j, _) in
            contains(grid: grid, sequence: ArraySlice(sequence), start: (i, j), impossible: &impossible)
        }
    }
}

private func contains(grid: [[Int]], sequence: ArraySlice<Int>, start: Cell, impossible: inout [[Set<Int>]]) -> Bool {
    if impossible[start.i][start.j].contains(sequence.count) {
        return false
    }

    if grid[start.i][start.j] == sequence.first! {
        // The first element matches, but what about the rest?
        let remaining = sequence.dropFirst()

        if remaining.isEmpty || neighbors(grid: grid, cell: start).contains { neighbor in
            contains(grid: grid, sequence: remaining, start: neighbor, impossible: &impossible)
        } {
            return true
        }
    }

    // Cache this failure.
    impossible[start.i][start.j].insert(sequence.count)
    return false
}

private func neighbors(grid: [[Int]], cell: Cell) -> [Cell] {
    return [
        (cell.i - 1, cell.j),
        (cell.i + 1, cell.j),
        (cell.i, cell.j - 1),
        (cell.i, cell.j + 1),
    ].filter { neighbor in
        neighbor.i >= 0 && neighbor.i < grid.count &&
        neighbor.j >= 0 && neighbor.j < grid[neighbor.i].count
    }
}

File: EPI/17.5/typescript/17.5.ts
type Cell = {i: number, j: number};

export default function contains(grid: number[][], seq: number[]): boolean {
    if (seq.length === 0) {
        return true;
    }

    // failed[i][j] contains k if there is no path starting at {i, j} that retraces the sequence
    // suffix starting at index k.
    const failed = grid.map(row =>
        row.map(() => new Set<number>())
    );

    return grid.some((row, i) =>
        row.some((_, j) => containsRec(grid, seq, 0, {i, j}, failed))
    );
}

function containsRec(grid: number[][], seq: number[], seqStart: number, {i, j}: Cell,
    failed: Set<number>[][]): boolean {

    if (failed[i][j].has(seqStart)) {
        return false;
    }

    if (grid[i][j] === seq[seqStart]) {
        // The first element matches, but what about the remaining elements?

        if (seqStart === seq.length - 1) {
            // There are no remaining elements.
            return true;
        }

        if (neighbors(grid, {i, j}).some(n => containsRec(grid, seq, seqStart + 1, n, failed))) {
            // At least one neighbor can retrace the remaining elements.
            return true;
        }
    }

    // Cache this failure.
    failed[i][j].add(seqStart);
    return false;
}

function neighbors(grid: number[][], {i, j}: Cell): Cell[] {
    return [
        {i: i - 1, j},
        {i: i + 1, j},
        {i, j: j - 1},
        {i, j: j + 1},
    ].filter(n =>
        n.i >= 0 && n.i < grid.length &&
        n.j >= 0 && n.j < grid[n.i].length
    );
}
